fix exec

papercuptech opened the pull request:
'exec' now handled directly in pyenv.bat
ExecCommand() in pyenv.vbs now uses WScript.Shell.exec()
  rather than creating 'exec.bat'

dedale commented:
I tried to rebase this dev 😅, but according to all the tests I added, this breaks ❌ `pyenv shell x.y.z` use case, which depends on exec.bat. 😢
Idk yet how to circumvent this problem. ⌛ ❓

dedale commented:
> 
> 
> I tried to rebase this dev 😅, but according to all the tests I added, this breaks ❌ `pyenv shell x.y.z` use case, which depends on exec.bat. 😢
> Idk yet how to circumvent this problem. ⌛ ❓

@kirankotari I think it also breaks `Plugin` feature

dedale commented:
Thanks @papercuptech, I think moving code to pyenv-shell.bat is the way to go.
To move on getting rid of exec.bat, I first need to add unit tests for `pyenv shell` and `Plugin` feature, then I'll rebase this PR.
I have already identified some impacts in the tests I have added.

papercuptech commented:
I've been working on fixes for the `shell` command and for plugins. 

I did move some logic to pyenv-shell.bat, but in order for the PYENV_VERSION var to be set correctly (and eliminate using `exec.bat`), `vbs` has to be avoided. 

Regarding plugins, even before my fix they suffer from the same argument mangling problem I originally fixed for `exec`, which was fixed by avoiding going through any `vbs` code. I've now fixed this issue for plugins as well. Note, my previous fix did break plugins in a different way, but that regression has also now been fixed

I've also rebased my changes, and I'm now working to run the unit tests.

Will update things in a day or two.

kirankotari commented:
@papercuptech @dedale let me change the base branch, that might help.

papercuptech commented:
Most tests passed. The first 2 in `test_pyenv` and all 4 in `test_pyenv_feature_shell` failed, but it looks to me like the problems are with the tests rather than what's being tested. My own "manual" testing indicates `shell` now works, plugins work, no `exec.bat` anywhere in sight. I attempted to debug tests, but not familiar with pytest and don't have the time atm to learn another test environment. Hopefully @dedale can address.

One "enhancement" is that plugins now run with PATH set just like it is with `exec` command; i.e. paths to "active" version of python.

dedale commented:
Great! You need to force Shell=false parameter where I call cmd.

On Wed, Apr 21, 2021, 14:15 papercuptech ***@***.***> wrote:

> I've been working on fixes for the shell command and for plugins.
>
> I did move some logic to pyenv-shell.bat, but in order for the
> PYENV_VERSION var to be set correctly (and eliminate using exec.bat), vbs
> has to be avoided.
>
> Regarding plugins, even before my fix they suffer from the same argument
> mangling problem I originally fixed for exec, which was fixed by avoiding
> going through any vbs code. I've now fixed this issue for plugins as
> well. Note, my previous fix did break plugins in a different way, but that
> regression has also now been fixed
>
> I've also rebased my changes, and I'm now working to run the unit tests.
>
> Will update things in a day or two.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyenv-win/pyenv-win/pull/140#issuecomment-824013816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZBCAHE3RDAKT2FGNKLBDTJ26X5ANCNFSM4PYULXZQ>
> .
>

dedale commented:
Commit to fix tests: https://github.com/dedale/pyenv-win/commit/ced501abfe1135656d2342174c199e4d43eae5f4

kirankotari commented:
> Commit to fix tests: [dedale@ced501a](https://github.com/dedale/pyenv-win/commit/ced501abfe1135656d2342174c199e4d43eae5f4)

@dedale are you able to push the changes to this PR?

dedale commented:
> 
> 
> > Commit to fix tests: [dedale@ced501a](https://github.com/dedale/pyenv-win/commit/ced501abfe1135656d2342174c199e4d43eae5f4)
> 
> @dedale are you able to push the changes to this PR?

I have created a PR to @papercuptech 's repo: https://github.com/papercuptech/pyenv-win/pull/1

papercuptech commented:
Merged @dedale changes into PR, but tests still failing..

dedale commented:
> Merged @dedale changes into PR, but tests still failing..

In Travis logs, pyenv.bat is getting `-m` argument. Did not reproduce on my laptop. I will investigate and use a temp PR to ensure tests pass in Travis.

dedale commented on pyenv-win/bin/pyenv.bat:
```diff
@@ -1,16 +1,93 @@
-IF EXIST "%~dp0"..\exec.bat (
-   call "%~dp0"..\exec.bat
+rem use pyenv.vbs to aid resolving absolute path of "active" version into 'bindir'
+for /f %%i in ('%pyenv% version') do call :normalizepath "%~dp0..\versions\%%i" bindir
```
```suggestion
for /f %%i in ('%pyenv% vname') do call :normalizepath "%~dp0..\versions\%%i" bindir
```
When no version is defined, `bindir` will be `...\versions\pyenv` because `pyenv version` returns:

> No global python version has been set yet. Please set the global version by typing:
> **pyenv** global 3.7.2

papercuptech replied:
A check was added for processing `exec` command, ensuring some version is active, otherwise "No global python version has been set yet. Please set the global version." is output

dedale commented on pyenv-win/bin/pyenv.bat:
```diff
@@ -1,16 +1,93 @@
+call :normalizepath %exe% exe
+
+if exist "%exe%.bat" (
+  set "exe=endlocal & call "%exe%.bat""
```
It seems this `endlocal` does not work properly.

Tests pass locally on my laptop because I use a `poetry` virtual env that updates the `PATH` so that when I run `pytest` it will execute the right `pytest.exe` directly.

In Travis CI, we use pyenv to run pytest. In the logs, it says that the arguments to pyenv start with `-m`.
This is because the call to `python -m pytest -v -s tests` actually uses `python.bat` in pyenv shims.
If calling `pyenv-shell.bat` in the unit tests still get `-m` as first argument, I guess this means `params` variable is not reset.

If I empty `params` variable before `:paramloop` (_but this is a dirty fix for me_), only one test fail: `test_shell_set_installed_version`. I guess this test fails for the same reason because `PYENV_VERSION` is not properly set because of `setlocal/endlocal` scopes.

papercuptech replied:
The test's expose an edge case when `pyenv.bat` indirectly calls back into itself; I would be surprised of this happens during normal usage.

Regardless, simply because of how environment variables are the only kind in '.bat' files, along with how setlocal/endlocal work, in the case of reentrancy, `params` and `bindir` need to be explicitly set to empty; so, not really a "dirty" fix, but the only fix.

I had previously stated that with this PR, `.bat` type plugins would have PATH prefixed with the "active" version path, but this was not true. In fact, it is not really possible to set PATH only for the duration of `exec` or the plugin, while still allowing the plugin to alter the calling shell's environment variables, which the `shell` command must be able to do.

I've special cased processing `.cmd` plugins, and the context for running things now is as follows:
* For the `exec` command, and for `.vbs`, and `.cmd` plugins, PATH is prefixed with the "active" python version, but none of them can alter any shell environment variables of the calling process/shell.
* For `.bat` plugins, PATH is **_not_** prefixed, but such plugins can modify the calling shell's environment variables, like pyenv-shell.bat does (as long they don't themselves call `setlocal`).

dedale commented on pyenv-win/bin/pyenv.bat:
```diff
@@ -1,16 +1,103 @@
+rem ====================================================================================
+:exec
+
+echo %bindir%
```
Should be removed, no?

papercuptech replied:
Yes it should, and also script should exit when no active version. Will fix shortly.

dedale commented on pyenv-win/bin/pyenv.bat:
```diff
@@ -1,16 +1,103 @@
+set "params="
+:paramloop
+if not [%2]==[] (
+  set params=%params% %2
```
```suggestion
  set "params=%params% %2"
```

papercuptech replied:
I can't recall the strange edge case I ran into, that led to always wrapping `set` arguments in double quotes, but it was likely related to spaces and/or % expansion and/or quoting behaving non-intuitively.

In this particular case wrapping has no effect, as the the passed parameters (probably better to say passed arguments, but I digress), have already been delimited by spaces. However, I will change for consistency.
